src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -0,0 +1,802 @@
+            _status = (ServiceControllerStatus)status.currentState;
+            _type = status.serviceType;
+            _statusGenerated = true;
+        }
+
+        /// Used by the GetServicesInGroup method.
+        private ServiceController(string machineName, Interop.ENUM_SERVICE_STATUS_PROCESS status)
+        {
+            if (!CheckMachineName(machineName))
+                throw new ArgumentException(SR.Format(SR.BadMachineName, machineName));
+
+            _machineName = machineName;
+            _name = status.serviceName;
+            _displayName = status.displayName;
+            _commandsAccepted = status.controlsAccepted;
+            _status = (ServiceControllerStatus)status.currentState;
+            _type = status.serviceType;
+            _statusGenerated = true;
+        }
+
+        /// Tells if the service referenced by this object can be paused.
+        public bool CanPauseAndContinue
+        {
+            get
+            {
+                GenerateStatus();
+                return (_commandsAccepted & Interop.ACCEPT_PAUSE_CONTINUE) != 0;
+            }
+        }
+
+
+        /// Tells if the service is notified when system shutdown occurs.
+        public bool CanShutdown
+        {
+            get
+            {
+                GenerateStatus();
+                return (_commandsAccepted & Interop.ACCEPT_SHUTDOWN) != 0;
+            }
+        }
+
+        /// Tells if the service referenced by this object can be stopped.
+        public bool CanStop
+        {
+            get
+            {
+                GenerateStatus();
+                return (_commandsAccepted & Interop.ACCEPT_STOP) != 0;
+            }
+        }
+
+        /// The descriptive name shown for this service in the Service applet.
+        public string DisplayName
+        {
+            get
+            {
+                if (_displayName == null)
+                    GenerateNames();
+                return _displayName;
+            }
+        }
+
+        /// The set of services that depend on this service. These are the services that will be stopped if
+        /// this service is stopped.
+        public ServiceController[] DependentServices
+        {
+            get
+            {
+                if (_dependentServices == null)
+                {
+                    IntPtr serviceHandle = GetServiceHandle(Interop.SERVICE_ENUMERATE_DEPENDENTS);
+                    try
+                    {
+                        // figure out how big a buffer we need to get the info
+                        int bytesNeeded = 0;
+                        int numEnumerated = 0;
+                        bool result = Interop.mincore.EnumDependentServices(serviceHandle, Interop.SERVICE_STATE_ALL, IntPtr.Zero, 0,
+                            ref bytesNeeded, ref numEnumerated);
+                        if (result)
+                        {
+                            _dependentServices = new ServiceController[0];
```

Comment:
I think it's going to be pretty common for a service not to have any dependent services.  I don't have any numbers from the wild, but I think it's worth making this Array.Empty().  I missed this when doing the initial port from the desktop sources.